### PR TITLE
[KTO]: Fix nan losses and crashing job

### LINKED ISSUE #1447: KTO training produces NaN rewards
Within the training with KTO Trainer I occasionally experience `nan` values as rewards.  
I am running the training as a job on Ms Azure with one GPU (NVIDIA A100 80GB PCIe).  
Ultimately these issues cause my Azure job to crash and retry...  

The log output I get from the KTOTrainer:
```
{'loss': 0.0202, 'grad_norm': 0.016206106171011925, 'learning_rate': 8.595282268751656e-06, 'rewards/chosen': 11.158143043518066, 'rewards/rejected': -29.0671443939209, 'rewards/margins': 40.22528839111328, 'kl': 0.0, 'logps/chosen': -15.192975044250488, 'logps/rejected': -180.1438446044922, 'epoch': 0.43}
{'loss': 0.0155, 'grad_norm': 4.091757774353027, 'learning_rate': 8.568778160614896e-06, 'rewards/chosen': 10.752923965454102, 'rewards/rejected': -26.606868743896484, 'rewards/margins': 37.35979461669922, 'kl': 0.0, 'logps/chosen': -13.974691390991211, 'logps/rejected': -156.9815673828125, 'epoch': 0.44}
{'loss': 0.0124, 'grad_norm': 0.06709074974060059, 'learning_rate': 8.542274052478135e-06, 'rewards/chosen': 10.838713645935059, 'rewards/rejected': -29.24416732788086, 'rewards/margins': 40.08287811279297, 'kl': 0.0, 'logps/chosen': -10.99155044555664, 'logps/rejected': -165.8121795654297, 'epoch': 0.44}
{'loss': 0.0113, 'grad_norm': 14.28693675994873, 'learning_rate': 8.515769944341374e-06, 'rewards/chosen': 11.07004451751709, 'rewards/rejected': -30.99440574645996, 'rewards/margins': 42.064453125, 'kl': 0.0, 'logps/chosen': -13.967004776000977, 'logps/rejected': -176.50094604492188, 'epoch': 0.45}
{'loss': 0.0193, 'grad_norm': 3.899095296859741, 'learning_rate': 8.489265836204611e-06, 'rewards/chosen': 10.825413703918457, 'rewards/rejected': -34.434303283691406, 'rewards/margins': 45.25971984863281, 'kl': 0.0, 'logps/chosen': -12.9598388671875, 'logps/rejected': -186.38381958007812, 'epoch': 0.46}
{'loss': 0.0109, 'grad_norm': 0.009407841600477695, 'learning_rate': 8.46276172806785e-06, 'rewards/chosen': nan, 'rewards/rejected': -33.95360565185547, 'rewards/margins': nan, 'kl': 0.0, 'logps/chosen': nan, 'logps/rejected': -176.4713897705078, 'epoch': 0.47}
{'loss': 0.0324, 'grad_norm': 17.832523345947266, 'learning_rate': 8.43625761993109e-06, 'rewards/chosen': 10.286358833312988, 'rewards/rejected': -33.60068893432617, 'rewards/margins': 43.887046813964844, 'kl': 0.0, 'logps/chosen': -20.224634170532227, 'logps/rejected': -184.18112182617188, 'epoch': 0.48}
{'loss': 0.0029, 'grad_norm': 0.03802444413304329, 'learning_rate': 8.409753511794329e-06, 'rewards/chosen': 10.086004257202148, 'rewards/rejected': nan, 'rewards/margins': nan, 'kl': 0.0, 'logps/chosen': -12.816671371459961, 'logps/rejected': nan, 'epoch': 0.48}
{'loss': 0.012, 'grad_norm': 2.815098524093628, 'learning_rate': 8.383249403657568e-06, 'rewards/chosen': 10.549690246582031, 'rewards/rejected': -31.304590225219727, 'rewards/margins': 41.85428237915039, 'kl': 0.0, 'logps/chosen': -13.178544998168945, 'logps/rejected': -169.447509765625, 'epoch': 0.49}
{'loss': 0.0074, 'grad_norm': 0.001768477726727724, 'learning_rate': 8.356745295520805e-06, 'rewards/chosen': 11.22235107421875, 'rewards/rejected': -33.09156799316406, 'rewards/margins': 44.31391906738281, 'kl': 0.0, 'logps/chosen': -13.94648265838623, 'logps/rejected': -178.08566284179688, 'epoch': 0.5}
{'loss': 0.0055, 'grad_norm': 8.117822647094727, 'learning_rate': 8.330241187384045e-06, 'rewards/chosen': 11.166982650756836, 'rewards/rejected': nan, 'rewards/margins': nan, 'kl': 0.0, 'logps/chosen': -14.707374572753906, 'logps/rejected': nan, 'epoch': 0.51}
{'loss': 0.0206, 'grad_norm': 1.6973105669021606, 'learning_rate': 8.303737079247284e-06, 'rewards/chosen': 10.326757431030273, 'rewards/rejected': -33.753868103027344, 'rewards/margins': 44.08062744140625, 'kl': 0.0, 'logps/chosen': -19.15297508239746, 'logps/rejected': -181.1234130859375, 'epoch': 0.52}
{'loss': 0.0136, 'grad_norm': 9.740607261657715, 'learning_rate': 8.277232971110523e-06, 'rewards/chosen': 10.298160552978516, 'rewards/rejected': nan, 'rewards/margins': nan, 'kl': 0.0, 'logps/chosen': -15.718103408813477, 'logps/rejected': nan, 'epoch': 0.52}
```

my pip freeze:
```
accelerate==0.28.0
aiohttp==3.9.3
aiosignal==1.3.1
async-timeout==4.0.3
attrs==23.2.0
bitsandbytes==0.43.0
certifi==2024.2.2
charset-normalizer==3.3.2
datasets==2.18.0
dill==0.3.8
docstring_parser==0.16
filelock==3.13.1
frozenlist==1.4.1
fsspec==2024.2.0
huggingface-hub==0.21.4
idna==3.6
Jinja2==3.1.3
markdown-it-py==3.0.0
MarkupSafe==2.1.5
mdurl==0.1.2
mpmath==1.3.0
multidict==6.0.5
multiprocess==0.70.16
networkx==3.2.1
numpy==1.26.4
nvidia-cublas-cu12==12.1.3.1
nvidia-cuda-cupti-cu12==12.1.105
nvidia-cuda-nvrtc-cu12==12.1.105
nvidia-cuda-runtime-cu12==12.1.105
nvidia-cudnn-cu12==8.9.2.26
nvidia-cufft-cu12==11.0.2.54
nvidia-curand-cu12==10.3.2.106
nvidia-cusolver-cu12==11.4.5.107
nvidia-cusparse-cu12==12.1.0.106
nvidia-nccl-cu12==2.19.3
nvidia-nvjitlink-cu12==12.4.99
nvidia-nvtx-cu12==12.1.105
packaging==24.0
pandas==2.2.1
peft==0.9.0
protobuf==5.26.0
psutil==5.9.8
pyarrow==15.0.1
pyarrow-hotfix==0.6
Pygments==2.17.2
python-dateutil==2.9.0.post0
pytz==2024.1
PyYAML==6.0.1
regex==2023.12.25
requests==2.31.0
rich==13.7.1
safetensors==0.4.2
sentencepiece==0.2.0
shtab==1.7.1
six==1.16.0
sympy==1.12
tokenizers==0.15.2
torch==2.2.1
tqdm==4.66.2
transformers==4.38.2
triton==2.2.0
trl @ git+https://github.com/huggingface/trl@a2aa0f0b09671eaf81a945eb5e4913165fee92fa
typing_extensions==4.10.0
tyro==0.7.3
tzdata==2024.1
urllib3==2.2.1
xxhash==3.4.1
yarl==1.9.4
```

the training script I use:
```python
from dataclasses import dataclass, field
from typing import Optional

from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer, HfArgumentParser, BitsAndBytesConfig

from trl import KTOConfig, KTOTrainer, ModelConfig
from peft import LoraConfig, TaskType, get_peft_model, prepare_model_for_kbit_training
import torch


# Define and parse arguments.
@dataclass
class ScriptArguments:
    """
    The arguments for the KTO training script.
    """

    dataset_path: Optional[str] = field(default=None, metadata={"help": "the online dataset to use, should include keys: [prompt, completion, label] OR [messages, completion, label]"})
    data_files: Optional[str] = field(default=None, metadata={"help": "the file(s) including data to use, this looks for 'data/{data_files}_train/test.jsonl.gz'. Datasets should include keys: [prompt, completion, label] OR [messages, completion, label]"})
    file_type: Optional[str] = field(default=None, metadata={"help": "the file type to open, e.g. 'json', 'csv'"})
    max_tokens: Optional[str] = field(default=4096, metadata={"help": "the maximum number of tokens returned by the data collator"})
    # debugging
    sanity_check: Optional[bool] = field(default=False, metadata={"help": "only train on 1000 samples"})
    


if __name__ == "__main__":
    parser = HfArgumentParser((ScriptArguments, KTOConfig, ModelConfig))
    script_args, kto_args, model_args = parser.parse_args_into_dataclasses()
    print(f"train with {script_args}, \n{model_args}")

    # Peft & Quantisation
    quantization_config = BitsAndBytesConfig(load_in_8bit=model_args.load_in_8bit)
    peft_config = LoraConfig(task_type=TaskType.CAUSAL_LM, r=model_args.lora_r, lora_alpha=model_args.lora_alpha, lora_dropout=model_args.lora_dropout)

    # Load the trainable model
    model = AutoModelForCausalLM.from_pretrained(model_args.model_name_or_path,
                                                 quantization_config = quantization_config,
                                                 trust_remote_code = None
                                                 torch_dtype = getattr(torch, model_args.torch_dtype) if model_args.torch_dtype != None else None,
                                                 device_map = "auto")#{'': 0} if torch.cuda.is_available() else "auto")

    model = prepare_model_for_kbit_training(model)
    model = get_peft_model(model, peft_config = peft_config)
    model.print_trainable_parameters()

    # Reference Model
    model_ref = AutoModelForCausalLM.from_pretrained(model_args.model_name_or_path,
                                                     quantization_config = quantization_config,
                                                     torch_dtype = getattr(torch, model_args.torch_dtype) if model_args.torch_dtype != None else None,
                                                     device_map = "auto")#{'': 0} if torch.cuda.is_available() else "auto")

    model_ref = prepare_model_for_kbit_training(model_ref)
    model_ref = get_peft_model(model_ref, peft_config=peft_config)

    # Load Tokenizer
    tokenizer = AutoTokenizer.from_pretrained(model_args.model_name_or_path)
    tokenizer.truncation_side = "left"
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token

    # Load the desired dataset
    if script_args.dataset_path != None:
        dataset = load_dataset(script_args.dataset_path)
    elif script_args.data_files != None and script_args.file_type != None:
        dataset = load_dataset(script_args.file_type, data_files={"train": f"./data/{script_args.data_files}_train.jsonl.gz", "test": f"./data/{script_args.data_files}_test.jsonl.gz"})    
    else:
        print("either dataset_path or data_files & file_type have to be defined")
        exit(1)

    if script_args.sanity_check == True:
        dataset["train"] = dataset["train"].select(range(1000))
    
    # Create Split if not existing already
    if "test" not in dataset:
        dataset = dataset["train"].train_test_split(train_size=0.9)
    
    # apply chat template if not preformatted
    if "prompt" not in dataset["train"].features:
        dataset = dataset.map(lambda x: {"prompt": tokenizer.apply_chat_template(x["messages"], tokenize=False, add_generation_prompt=False)})

    # Set max. lengths for DefaultDataCollator

    max_prompt_len, max_compl_len, max_len = 0, 0, 0
    tokenizer.model_max_length = script_args.max_tokens
    tokenizer.max_model_input_sizes = script_args.max_tokens

    for sample in dataset["train"]:

        compl_len = len(tokenizer(sample["completion"], truncation=True)["input_ids"])
        total_len = len(tokenizer(sample["prompt"] + sample["completion"], truncation=True)["input_ids"])
        prompt_len = total_len - compl_len
    
        max_prompt_len = max(max_prompt_len, prompt_len)
        max_compl_len = max(max_compl_len, compl_len)
        max_len = max(max_len, total_len)

    kto_args.max_prompt_length = max_prompt_len
    kto_args.max_completion_length = max_compl_len
    kto_args.max_length = max_len

    print(dataset)
    print(f"max_prompt_length={kto_args.max_prompt_length}, max_completion_length={kto_args.max_completion_length}, max_len={kto_args.max_length}")

    # set desired/undesired weights

    desired_weight = len(dataset['train']) / (2 * len(dataset["train"].filter(lambda d: d["label"] == True)))
    undesired_weight = len(dataset['train']) / (2 * len(dataset["train"].filter(lambda d: d["label"] == False)))

    kto_args.desirable_weight = desired_weight
    kto_args.undesirable_weight = undesired_weight

    # initialize the KTO trainer
    kto_trainer = KTOTrainer(
        model,
        model_ref,
        args=kto_args,
        train_dataset=dataset["train"],
        eval_dataset=dataset["test"],
        tokenizer=tokenizer
    )

    # train
    kto_trainer.train()
```

the call arguments
```
python train_kto.py \
    --model_name_or_path DiscoResearch/DiscoLM_German_7b_v1 \
    --data_files wp_rag_kto_20k \
    --file_type json \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 8 \
    --num_train_epochs 3 \
    --learning_rate 1e-5 \
    --gradient_accumulation_steps 2 \
    --logging_steps 10 \
    --evaluation_strategy epoch \
    --output_dir kto_finetuned \
    --optim adamw_bnb_8bit \
    --warmup_steps 10 \
    --logging_first_step \
    --use_peft \
    --lora_r 8 \
    --lora_alpha 16 \
    --report_to none \
    --disable_tqdm False \
    --beta 0.5 \
    --torch_dtype bfloat16 \
    --bf16 \
    --load_in_8bit
```

Maybe @lewtun can help

### PR DESCRIPTION
**Pull request by claralp**
fixes #1447   
- use `nanmean()` instead of `mean()` for losses to avoid nan losses
- remove obsolete `accelerator.gather` for metrics as the metrics are all collected to cpu anyway and averaged later when logging them.

### CODE REVIEW

**Comment by kashif** [comment]
thanks @claralp

**Comment by claralp** [comment]
btw. I also tested it on multiple GPUs now. It is running and both GPUs show usage, but the distribution does not seem to be effective as there is no speedup compared to 1 GPU.  
However, this is the same behavior as before this PR. Might look into that later

**Comment by claralp** [comment]
@kashif @lewtun: is there anything more I should validate or can you approve this for now?  
I am looking further into the multi GPU case atm, but would prefer to keep this small and open another PR if I find something.

**Comment by kashif** [comment]
@claralp would you be able to test out #1476 ?

**Comment by claralp** [comment]
@kashif for me https://github.com/huggingface/trl/pull/1476 has a few issues, (see comments there).  
Except from this:
- removing the interleaving of datasets works for me and is a useful change
- using e.g. `metrics[f"{prefix}rewards/chosen"] = self.accelerator.gather(mean_chosen_reward).nanmean().item()`
instead of `metrics[f"{prefix}rewards/chosen"] = chosen_rewards.detach().nanmean().cpu()` makes no difference for balanced batches. Both move the mean value to CPU at the end (one as tensor and one as scalar tho). 
However, `self.accelerator.gather(mean_chosen_reward)` leads to issues when `mean_chosen_reward` is nan. This happened if there are no 'chosen' samples in the batch

**Comment by kashif** [comment]
thanks @claralp with regard to moving the loss to the `self.args.device` I thought that `accelerate` handles all that for us?

**Comment by claralp** [comment]
@kashif for the loss `accelerate` handles that for us, yes.  
But this is just for the rewards/logp that are stored as metrics to be logged. In my understanding, that's why the metrics are collected across all GPUs and sent to CPU to calculate the mean value there.  
E.g. in dpo_trainer, you use `metrics[f"{prefix}rewards/chosen"] = chosen_rewards.mean().cpu()`

File: trl/trainer/kto_trainer.py [hunk 1]
```diff
@@ -1027,6 +1008,9 @@ def compute_loss(
 
         loss, metrics = self.get_batch_loss_metrics(model, inputs, train_eval="train")
 
+        # Make sure to move the loss to the device the original accumulating loss is at back in the `Trainer` class:
+        loss = loss.to(self.args.device)
```
Comment (claralp) [inline trl/trainer/kto_trainer.py:1012]: @kashif if your question was regarding this line:  
Without this I get an exception when running on multiple GPUs: 
`ValueError: Calculated loss must be on the original device: cuda:0 but device in use is cuda:1`.  
I copied this line from dpo_trainer to fix this, see [dpo_trainer.py#L1084](https://github.com/huggingface/trl/blob/main/trl/trainer/dpo_trainer.py#L1084)

**Review by kashif — APPROVED** [review]

**Comment by claralp** [comment]
@kashif, I applied the suggested changes from the code style pipeline. Should be green again now

**Comment by johncordeiro** [comment]
we're facing a problem that possibly will be solved by this PR, thanks for this @claralp ! we're going to test it and give the feedback.

**Comment by johncordeiro** [comment]
hey! I just tested the code in branch `claralp:fix_nans`, and it's hanging during evaluation.

**Comment by kawine** [comment]
@johncordeiro could you try the version here and see if you're sill experiencing hanging? https://github.com/kawine/trl if so, more context would be helpful

@claralp thank you for all the fixes! seems like this version is only using metrics from the main process' microbatch, by getting rid of the gathers? i have a pr in #1499 that includes most of your changes here but still supports batch-wise gathering if you want to take a look

**Comment by claralp** [comment]
> @johncordeiro could you try the version here and see if you're sill experiencing hanging? https://github.com/kawine/trl if so, more context would be helpful
> 
> @claralp thank you for all the fixes! seems like this version is only using metrics from the main process' microbatch, by getting rid of the gathers? i have a pr in #1499 that includes most of your changes here but still supports batch-wise gathering if you want to take a look

@kawine as I understand and experienced it, this version does not use only the main process' micro batch.
It sends the micro batch of every GPU directly to the CPU to calculate metrics there.  
Using `accelerator.gather` would only gather them all in the main process first before sending them to CPU. This works, but is not necessary, as the metrics are collected on CPU later anyway.

For DPO it is implemented the same without gathering: https://github.com/huggingface/trl/blob/0ee349dcd43b0f4b3169449f16751c38ac4a609f/trl/trainer/dpo_trainer.py#L1055

**Comment by kawine** [comment]
> It sends the micro batch of every GPU directly to the CPU to calculate metrics there.

thanks @claralp ! maybe it's just me, but I'm not seeing the CPU collection happen? if there is no explicit gather, then the logging only reflects the statistics from the microbatch on the main process only. 

[the HF documentation for Trainer](https://huggingface.co/docs/transformers/en/main_classes/trainer#transformers.Trainer.log_metrics), which is subclassed here, also says for `log_metrics` that "Under distributed environment this is done only for a process with rank 0."

**Comment by claralp** [comment]
@johncordeiro do you have `prediction_loss_only` enabled in your evaluation step? The logits were not propagated, so this part was not working in the previous version.  
If this is not the reason behind the hanging issue in evaluation, can you provide the line where it hangs?

**Comment by claralp** [comment]
@kawine it uses `.tolist()` for all metrics, which moves them to CPU. The `log_metrics` of HF Trainer is not used by KTO Trainer, it uses its own subclassed `log` function

**Comment by kawine** [comment]
@claralp if i add the line `metrics[f"device"] = torch.Tensor([float(str(self.args.device)[-1])]).cpu()` to `get_batch_loss_metrics`, i can see in wandb that the value is always 0 (i.e., the main process), suggesting that only metrics from one device are tracked without the explicit gather

**Comment by claralp** [comment]
> @claralp if i add the line `metrics[f"device"] = torch.Tensor([float(str(self.args.device)[-1])]).cpu()` to `get_batch_loss_metrics`, i can see in wandb that the value is always 0 (i.e., the main process), suggesting that only metrics from one device are tracked without the explicit gather

@kawine you need to read `self.accelerator.process_index` to get the current index when using Accelerator.  
And make sure you use the latest version, in a previous version `store_metrics` was really just called from the main process, which was wrong

File: trl/trainer/kto_trainer.py [hunk 2]
```diff
@@ -1111,25 +1098,19 @@ def prediction_step(
 
         prediction_context_manager = torch.cuda.amp.autocast if self._peft_has_been_casted_to_bf16 else nullcontext
         with torch.no_grad(), prediction_context_manager():
-            loss, metrics = self.get_batch_loss_metrics(model, inputs, train_eval="eval")
+            if prediction_loss_only:
+                loss, metrics = self.get_batch_loss_metrics(model, inputs)
+            else:
+                loss, metrics, logits = self.get_batch_loss_metrics(model, inputs, return_logits=True)
+                print(logits.shape)
```
Comment (alaeddine-13) [inline trl/trainer/kto_trainer.py:1105]:
```suggestion
```

**Comment by claralp** [comment]
> still getting the same thing with `metrics[f"device"] = torch.Tensor([float(str(self.accelerator.process_index)[-1])]).cpu()` and the latest version of accelerate (0.28.0)

@kawine, I just tested this with `per_device_batch_size=2` on 2 GPUs. When printing the metrics in `store_metrics`, I get:
```
{'rewards/chosen': [-0.23836135864257812, -0.9353666305541992], 'rewards/rejected': [], 'kl': 1.9421863555908203, 'logps/chosen': [-33.11909484863281, -32.70828628540039], 'logps/rejected': [], 'device': 1}
{'rewards/chosen': [-0.018026351928710938], 'rewards/rejected': [-0.438812255859375], 'kl': 1.9421863555908203, 'logps/chosen': [-24.179973602294922], 'logps/rejected': [-138.9338836669922], 'device': 0}
```
I print them in `store_metrics`, because the logging to e.g. wandb prints an average of 0.5 for 2 devices instead of showing the metrics per device.  
Which accelerate config do you use?

**Comment by claralp** [comment]
@kashif or @lewtun can you confirm if the current changes work for you as well?

**Comment by kashif** [comment]
sure @claralp let me see!

**Comment by kawine** [comment]
@claralp i've tried this with deepspeed and the regular data parallel (examples/accelerate_configs/multi_gpu.yaml) and it still only reports stats from the main process.

if you are printing the stats in store_metrics, it will print them from each process, yes, so you will have one line with `'device': 1` and another with `'device': 0`. But if you look at the chart for `train/device` in wandb, it will show only 0, because the two processes' stats are never gathered (at least for me)
Screen Shot 2024-04-03 at 1 40 13 PM

if you look at PPOTrainer, the metrics are explicitly being gathered across all machines at each step with an all_reduce: https://github.com/huggingface/trl/blob/ab0d11d81550618546cd1d8807627a594a58f029/trl/trainer/ppo_trainer.py#L897

**Comment by PhilipMay** [comment]
@kawine reading your plot -> is it possible that you train on multiple different machines with one GPU each?
It reads "copper-paper" and "noble-pyramid". I think thats names coming from k8s?

**Comment by kashif** [comment]
closing this due to #1514

**Pull request closed without merging**